controllers/apps/operations/expose.go

```
@@ -190,3 +222,180 @@ func (e ExposeOpsHandler) SaveLastConfiguration(reqCtx intctrlutil.RequestCtx, c
+			clusterDef, err := getClusterDefByName(reqCtx.Ctx, cli, cluster.Spec.ClusterDefRef)
+			if err != nil {
+				return nil, err
+			}
+			clusterCompDef := clusterDef.GetComponentDefByName(clusterCompSpec.ComponentDefRef)
+			if clusterCompDef == nil || clusterCompDef.Service == nil {
+				return nil, fmt.Errorf("referenced cluster component definition or services is not defined: %s", clusterCompSpec.ComponentDefRef)
+			}
+			return clusterCompDef.Service.ToSVCPorts(), nil
+		}
+		return nil, fmt.Errorf("component definition is not defined, cluster: %s, component: %s", cluster.Name, clusterCompSpec.Name)
+	}
+
+	defaultRoleSelectorFunc := func() (string, error) {
+		if clusterCompSpec.ComponentDef != "" {
+			compDef, err := component.GetCompDefinition(reqCtx, cli, cluster, clusterCompSpec.Name)
+			if err != nil {
+				return "", err
+			}
+			if len(compDef.Spec.Roles) == 0 {
+				return "", nil
+			}
+			for _, role := range compDef.Spec.Roles {
+				if role.Writable && role.Serviceable {
+					return role.Name, nil
+				}
+			}
+			return "", nil
+		}
+		if cluster.Spec.ClusterDefRef != "" && clusterCompSpec.ComponentDefRef != "" {
+			clusterDef, err := getClusterDefByName(reqCtx.Ctx, cli, cluster.Spec.ClusterDefRef)
+			if err != nil {
+				return "", err
+			}
+			clusterCompDef := clusterDef.GetComponentDefByName(clusterCompSpec.ComponentDefRef)
+			if clusterCompDef == nil {
+				return "", fmt.Errorf("referenced cluster component definition is not defined: %s", clusterCompSpec.ComponentDefRef)
+			}
+			switch clusterCompDef.WorkloadType {
+			case appsv1alpha1.Replication:
+				return constant.Primary, nil
+			case appsv1alpha1.Consensus:
+				return constant.Leader, nil
+			}
+		}
+		return "", nil
+	}
+
+	for _, exposeService := range exposeServices {
+		if checkServiceExist(exposeService) {
+			return fmt.Errorf("service %s already exists, cluster: %s, component: %s", exposeService.Name, cluster.Name, clusterCompSpec.Name)
```

According to the idempotent processing, if the service already exists, there is no need to report an error

---

> According to the idempotent processing, if the service already exists, there is no need to report an error

fixed done